docs/source/rag/retrieval_question_generation.rst

```diff
@@ -0,0 +1,208 @@
+
+Naturally, we want to be able to evaluate this retriever system for the RAG model to compare and judge its
+performance. To evaluate a retriever system, we would first need a test set of questions on the documents.
+These questions need to be diverse, relevant, and coherent. Manually generating questions may be challenging
+because it first requires you to understand the documents, and spend lots of time coming up with questions 
+for them. 
+
+We want to make this process simpler by utilizing an LLM to generate questions for this test set. This
+tutorial will walk through how to generate the questions and how to analyze the diversity and relevance
+of the questions.
+
+.. _retrieval-question-generation-quickstart:
+
+The following guide will walk you through generating the question test set for retriever evaluation.
+
+Step 1: Install and Load Packages
+==================================
+
+.. code-block:: python
+   import os
+   import pickle
+   from langchain.docstore.document import Document
+   from langchain.embeddings import OpenAIEmbeddings
+   import openai
+
+Step 2: Load Data
+==================
+Load the data through Langchain Documents to utilize their embedding models.
+
+.. code-block:: python
+   from langchain.docstore.document import Document
+   import pickle
+
+   docs, metadatas = None, None
+   with open("/dbfs/bbqiu/mlflow_docs.pkl", "rb") as f:
+   docs = pickle.load(f)
+   with open("/dbfs/bbqiu/mlflow_docs_metadata.pkl", "rb") as f:
+   metadatas = pickle.load(f)
+
+   documents = []
+   for i in range(len(docs)):
+   doc, mdata = docs[i], metadatas[i]
+   documents.append(Document(page_content=doc, metadata=mdata))
+
+   print(documents[0])
+
+Step 3: Set OpenAI Key
+======================
+Set the OpenAI key to utilize embeddings and GPT.
+
+.. code-block:: python
+   openai.api_key = "<redacted>"
+
+Step 4: Generate Questions
+==========================
+Generate the list of questions with GPT, utilizing prompt engineering to produce better quality questions.
+We have found that better quality questions are produced if you ask it for multiple questions per chunk.
+In addition, sometimes the question references the document without explaining what it is referencing.
+Telling GPT explicitly to explain references helps with this.
+
+.. code-block:: python
+   question_per_chunk = 5
+   queries = []
+   chunks = []
+   for doc in documents:
+      chunk = doc.page_content
+      chunks.append(chunk)
+      params = {
+         "model": "gpt-3.5-turbo",
+         "messages": [
+               {
+                  "role": "user",
+                  "content": f"{chunk}.\n Please generate {question_per_chunk} questions based on the above document. The questions should be diverse and ask for different aspects of the document. Don't give vague references to the document without description. Split each question with a newline"
+               }
+         ],
+      }
```

Comment:
Where did we get this prompt / how well does it work?

Reply:
I basically tried out some prompts @bbqiu had and made some changes, and this prompt was the best one so far. The part where it asks it to generate multiple questions rather than one gave significant improvement.